src/logging/metrics-logger.ts

```diff
@@ -0,0 +1,117 @@
+import {hrtime} from 'process';
+import {Event} from '../event.js';
+import {DefaultLogger} from './default-logger.js';
+
+interface Metric {
+  name: string;
+  matches: (event: Event) => boolean;
+  count: number;
+}
+
+/**
+ * A {@link Logger} that keeps track of metrics.
+ */
+export class MetricsLogger extends DefaultLogger {
+  private _startTime: [number, number] = hrtime();
+  private readonly _metrics: Metric[] = [
+    {
+      name: 'Success',
+      // 'no-command' is technically a success, but we don't want to count it as
+      // a success for this metric because nothing was actually run.
+      matches: (e: Event) => e.type === 'success' && e.reason !== 'no-command',
+      count: 0,
+    },
+    {
+      name: 'Ran',
+      matches: (e: Event) => e.type === 'success' && e.reason === 'exit-zero',
+      count: 0,
+    },
+    {
+      name: 'Skipped (fresh)',
+      matches: (e: Event) => e.type === 'success' && e.reason === 'fresh',
+      count: 0,
+    },
+    {
+      name: 'Restored from cache',
+      matches: (e: Event) => e.type === 'success' && e.reason === 'cached',
+      count: 0,
+    },
+  ];
+
+  /**
+   * @param rootPackage The npm package directory that the root script being
+   * executed belongs to.
+   */
+  constructor(rootPackage: string) {
+    super(rootPackage);
+  }
+
+  /**
+   * Update relevant metrics for an event and pass it up to the parent logger.
+   */
+  override log(event: Event): void {
+    this._updateMetrics(event);
+    super.log(event);
+  }
+
+  /**
+   * Log the current metrics and reset the state of each metric.
+   */
+  override printMetrics(): void {
+    const successes = this._metrics[0].count;
+
+    if (!successes) {
+      this._resetMetrics();
+      return;
+    }
+
+    const elapsed = this._getElapsedTime();
+    const nameOffset = 20;
+
+    const out: string[] = [
+      `🏁 [metrics] Executed ${successes} script(s) in ${elapsed} seconds`,
+    ];
+
+    for (const metric of this._metrics.slice(1)) {
+      const name = metric.name.padEnd(nameOffset);
+      const count = metric.count;
+      const percent = this._calculatePercentage(count, successes);
+
+      out.push(`\t${name}: ${count} (${percent}%)`);
+    }
+
+    console.log(out.join('\n'));
+
+    this._resetMetrics();
+  }
+
+  private _updateMetrics(event: Event): void {
+    for (const metric of this._metrics) {
+      if (metric.matches(event)) {
+        metric.count++;
+      }
+    }
+  }
+
+  private _resetMetrics(): void {
+    this._startTime = hrtime();
```

Comment:
This does the wrong thing in watch mode I think. We end up reporting the time from the end of the last run, rather than the beginning of this run.

So e.g. if you go get coffee for five minutes, then come back and trigger a build that takes three seconds, the elapsed time will be reported as five minutes and three seconds

Reply:
Nice catch. I just added some code to reset the metrics when a `watch-run-start` event is received, I assume this is when we would want to 'start the timer'?

Reply:
Yes I think this is correct now :+1: